src/half_bridge_tim.cpp

```diff
@@ -0,0 +1,283 @@
+/* LibreSolar charge controller firmware
+ * Copyright (c) 2016-2019 Martin Jäger (www.libre.solar)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef UNIT_TEST
+
+#include "half_bridge.h"
+#include "pcb.h"
+#include "config.h"
+
+
+static int _pwm_resolution;
+static float _min_duty;
+static float _max_duty;
+static uint8_t _deadtime_clocks;
+
+
+static bool _enabled;
+
+class PWM_TIM3
+{
+    public:
+
+    static void _init_registers(uint32_t pwm_resolution)
+    {
+        // Enable peripheral clock of GPIOB
+#if defined(STM32F0)
```

Comment:
The STM32F0 always uses TIM1, which is more suitable for PWM generation (includes automatic dead-time generation etc.). Only the L0 uses TIM3, as it doesn't have TIM1.

Reply:
TIM1/F0 vs. TIM3/L0 see below.
Classes vs. Namespaces: Classes are more or less exactly meant for that: bundling a couple of functions on a data structure.

In a nutshell: For the moment: It would not make much of a difference.
And I agree, using name spaces would keep the code fully "procedural".

However, I had something in mind when I used classes. If we ever want to have 2 half_bridges controlled, we can go the OO way from here easily, we just make a half_bridge class and pass the low-level access class as a template parameter for instance. Or implement the high-level half-bridge interface (half_bridge_...() functions) as a class half_bridge with virtual functions for the low level access (what is now in the PWM_TIMx  classes). And real C++ experts would come up probably with a dozen other ways to achieve the same.    

In an incremental approach, however, I just solved the problem at hand (the code duplication) and did not change the external interface (but had that in mind). Namespaces would not work for that use case, so I did not consider them.

Reply:
Yes -> Unit Test Mockup -> Very simply to do (if you know what you are doing). BTW, I would like to understand how to run the unit testing, is there any documentation on this?

Reply:
In platformio it's as simple as running
```
platformio test -e unit-test-native
```
More documentation can be found [here](https://docs.platformio.org/en/latest/plus/unit-testing.html).

Unfortunately, only pressing the test button in VS code doesn't work as it tries to run the test also for "normal" build environments and not only the unit-test-native as defined in platformio.ini. Or the other ones have to be commented out.

Reply:
Ok, thanks. Had to install MinGW and to add a -D __STDC_FORMAT_MACROS in platformio.ini for build_flags in env:unit-test-native and "my" tests are passing now.

Reply:
I did add some unit tests, and it even helped me to find a bug. I am not sure if I did it in the way you had in mind. I put the PWM_HW mockup class in the same file as the normal code. 

Factoring out the stub would of course be possible, but at least the SystemCoreClock const has to remain. Unless we do some more changes. I would suggest to pass the SystemCoreClock as argument to the half_bride_init function, this way we avoid referring to a global const which does not exist in the UNIT TEST environment.